### PR TITLE
Add ability to switch between voice & video in calls

### PR DESCRIPTION
**Pull request by SimonBrandner**
Type: enhancement
Requires https://github.com/matrix-org/matrix-js-sdk/pull/2024
Fixes https://github.com/vector-im/element-web/issues/18619

<hr>

Screenshot_20211117_145353
Screenshot_20211117_145422

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add ability to switch between voice & video in calls ([\#7155](https://github.com/matrix-org/matrix-react-sdk/pull/7155)). Fixes vector-im/element-web#18619. Contributed by @SimonBrandner.

Preview: https://61e44e908c22dc189d9b3ddb--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/components/views/voip/CallView.tsx [hunk 1]
```diff
@@ -347,35 +347,38 @@ export default class CallView extends React.Component<IProps, IState> {
     };
 
     private renderCallControls(): JSX.Element {
-        // We don't support call upgrades (yet) so hide the video mute button in voice calls
-        const vidMuteButtonShown = this.props.call.type === CallType.Video;
+        const { call, pipMode } = this.props;
+        const { primaryFeed, callState, micMuted, vidMuted, screensharing, sidebarShown } = this.state;
+
+        // If SDPStreamMetadata isn't supported don't show video mute button in voice calls
+        const vidMuteButtonShown = call.opponentSupportsSDPStreamMetadata() || call.hasLocalUserMediaVideoTrack;
```
Comment (t3chguy) [inline src/components/views/voip/CallView.tsx:354]: Any reason we went from `call.type === CallType.Video` to `call.hasLocalUserMediaVideoTrack` ?

File: src/components/views/voip/CallView.tsx [hunk 2]
```diff
@@ -347,35 +347,38 @@ export default class CallView extends React.Component<IProps, IState> {
...
         // Screensharing is possible, if we can send a second stream and
         // identify it using SDPStreamMetadata or if we can replace the already
         // existing usermedia track by a screensharing track. We also need to be
         // connected to know the state of the other side
         const screensharingButtonShown = (
-            (this.props.call.opponentSupportsSDPStreamMetadata() || this.props.call.type === CallType.Video) &&
-            this.props.call.state === CallState.Connected
+            (call.opponentSupportsSDPStreamMetadata() || call.hasLocalUserMediaVideoTrack) &&
```
Comment (t3chguy) [inline src/components/views/voip/CallView.tsx:360]: this change seems would seem to hide the screen share button if you have no local webcams? AIUI we support asymmetrical video calls now so this should be fine the way it was?

File: src/components/views/voip/CallView.tsx [hunk 3]
```diff
@@ -482,7 +485,7 @@ export default class CallView extends React.Component<IProps, IState> {
 
         // This is a bit messy. I can't see a reason to have two onHold/transfer screens
         if (isOnHold || transfereeCall) {
-            if (isVideoCall) {
+            if (call.hasLocalUserMediaVideoTrack || call.hasRemoteUserMediaVideoTrack) {
```
Comment (t3chguy) [inline src/components/views/voip/CallView.tsx:488]: this looks like the perfect place to use `call.type === CallType.Video` for readability

**Review by t3chguy — COMMENTED** [review]
Looks sane otherwise

File: src/components/views/voip/CallView.tsx (hunk 1 above)
Comment (SimonBrandner) [reply src/components/views/voip/CallView.tsx:354]: Call types are a weird thing that don't really work well once you have a call where one side is sending video and the other isn't. Therefore we now have `hasLocalUserMediaVideoTrack` and `hasRemoteUserMediaVideoTrack` which more accurately represent the call's state

File: src/components/views/voip/CallView.tsx (hunk 2 above)
Comment (SimonBrandner) [reply src/components/views/voip/CallView.tsx:360]: If you have no webcam and want to screen share you need the opponent to support SDPStreamMetadata (you can send a second stream), if it doesn't do that you need to be already sending a video track which you can replace.

**Review by t3chguy — APPROVED** [review]

**Review by gaelledel — APPROVED** [review]
Thank you!!

**Pull request merged**